execution/binary/vector.go

```diff
@@ -104,84 +105,24 @@ func (o *vectorOperator) Explain() (me string, next []model.VectorOperator) {
 }
 
 func (o *vectorOperator) Series(ctx context.Context) ([]labels.Labels, error) {
-	var err error
-	o.once.Do(func() { err = o.initOutputs(ctx) })
-	if err != nil {
+	if err := o.initOnce(ctx); err != nil {
 		return nil, err
 	}
-
 	return o.series, nil
 }
 
-func (o *vectorOperator) initOutputs(ctx context.Context) error {
-	var highCardSide []labels.Labels
-	var errChan = make(chan error, 1)
-	go func() {
-		var err error
-		highCardSide, err = o.lhs.Series(ctx)
-		if err != nil {
-			errChan <- err
-		}
-		close(errChan)
-	}()
-
-	lowCardSide, err := o.rhs.Series(ctx)
-	if err != nil {
-		return err
-	}
-	if err := <-errChan; err != nil {
-		return err
-	}
-
-	o.lhSampleIDs = highCardSide
-	o.rhSampleIDs = lowCardSide
-
-	if o.matching.Card == parser.CardOneToMany {
-		highCardSide, lowCardSide = lowCardSide, highCardSide
-	}
-
-	buf := make([]byte, 1024)
-	var includeLabels []string
-	if len(o.matching.Include) > 0 {
-		includeLabels = o.matching.Include
-	}
-	keepLabels := o.matching.Card != parser.CardOneToOne
-	keepName := !shouldDropMetricName(o.opType, o.returnBool)
-	highCardHashes, highCardInputMap := o.hashSeries(highCardSide, keepLabels, keepName, buf)
-	lowCardHashes, lowCardInputMap := o.hashSeries(lowCardSide, keepLabels, keepName, buf)
-	output, highCardOutputIndex, lowCardOutputIndex := o.join(highCardHashes, highCardInputMap, lowCardHashes, lowCardInputMap, includeLabels)
-
-	series := make([]labels.Labels, len(output))
-	for _, s := range output {
-		series[s.ID] = s.Metric
-	}
-	o.series = series
-
-	o.outputCache = make([]outputSample, len(series))
-	for i := range o.outputCache {
-		o.outputCache[i].lhT = -1
-	}
-	o.pool.SetStepSize(len(highCardSide))
-
-	o.table = newTable(
-		o.pool,
-		o.matching.Card,
-		o.operation,
-		o.outputCache,
-		newHighCardIndex(highCardOutputIndex),
-		lowCardinalityIndex(lowCardOutputIndex),
-	)
-
-	return nil
-}
-
 func (o *vectorOperator) Next(ctx context.Context) ([]model.StepVector, error) {
 	select {
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	default:
 	}
 
+	// some operators do not call Series of all their children
```

Comment:
I wonder if we want to make this a contract in the future. It would help avoid some gotchas.

```suggestion
	// Some operators do not call Series of all their children.
```

Reply:
:heavy_check_mark: